src/core/experimental/memutils.d

```diff
@@ -0,0 +1,179 @@
+/**
+ * Pure D replacement of the C Standard Library basic memory building blocks of string.h
+ * Source: $(DRUNTIMESRC core/experimental/memutils.d)
+ */
+module core.experimental.memutils;
+
+/** memset() implementation */
+
+/**
+ * NOTE(stefanos):
+ * Range-checking is not needed since the user never
+ * pass an `n` (byte count) directly.
+ */
+
+/*
+  If T is an array,set all `dst`'s bytes
+  (whose count is the length of the array times
+  the size of the array element) to `val`.
+  Otherwise, set T.sizeof bytes to `val` starting from the address of `dst`.
+*/
+
+void memset(T)(ref T dst, const ubyte val)
+{
+    import core.internal.traits : isArray;
+    const uint v = cast(uint) val;
+    static if (isArray!T)
+    {
+        size_t n = dst.length * typeof(dst[0]).sizeof;
+        Dmemset(dst.ptr, v, n);
+    }
+    else
+    {
+        Dmemset(&dst, v, T.sizeof);
+    }
+}
+
+version (GNU)
```

Comment:
AFAIK GDC does not set the `D_SIMD` version, so you probably don't need the explicit GNU branch here.

Reply:
Indeed it doesn't.  `D_SIMD` is to be defined when you have `__simd` as intrinsics.

Reply:
Your comment revealed an important let's say "bug". LDC doesn't set `D_SIMD` as well, which means that LDC uses the naive version as well. Although that reaches parity with libc, using the actual SIMD version surpasses libc quite a lot, especially for small sizes. And I don't know a way to check if SIMD is available in LDC. Is there such a thing in GDC?

Reply:
GDC pretty much supports any kind of vector size. `__traits(compiles, the vector you want to use)` is the best way to detect if it is rejected for unsupported reasons.